### PR TITLE
rotate ROIs

### PR DESCRIPTION
**Pull request by stropitek**
- **wip: rotate add roi rotation**
- **wip: normalize and denormalize with different rotation centers**
- **wip!: bound roi takes angle into account**
- **wip: partial implementation of resize with angle but not satisfying**
- **wip: resize rotated rois**
- **fix: better handler sizes for small rois**
- **fix: do not implement bounding rules if rotated for now**
- **fix: make it harder to misclick outside of the rotate handler**
- **chore: do not implement rotate after callback**
- **feat: activate rotate under prop**
- **fix: adapt resize cursors to rotation angle**
- **fix: make sure initial roi angle is in the correct range**
- **fix: correct rotate with overlay and improve rotate handler size**
- **fix: correct size of box shadow**

### CODE REVIEW

**Comment by stropitek** [comment]
@targos can you have a look and see if you spot any bug in the stories?

You can look at the "rotate" and the "shadow" stories

**Comment by stropitek** [comment]
PS: for now rotated ROIs are not moved / resized to fit inside the image if an action puts them partially out of it. It's not easy to implement and not easy to decide what's the best way to do it. When it happens, the action is simply canceled. We don't really need this feature anyway.

**Comment by targos** [comment]
1. I would not show the rotation handle while the ROI is being created:

CleanShot 2024-03-04 at 08 41 50

2. I'm able to make it go out of the boundaries:

CleanShot 2024-03-04 at 08 45 02

3. The rotate handle can be brought outside of the boundaries, making it impossible to see/use. Maybe we can accept this limitation.

CleanShot 2024-03-04 at 08 46 52

4. Did you consider drawing a line between the edge of the rectangle and the circle?

File: src/utilities/corners.ts [hunk 1]
```diff
@@ -12,64 +12,80 @@ export interface CornerData {
   cy: number;
 }
 
+// Expects angle input ranging from -PI to +PI
+function getCursorIndex(angle: number, idx: number) {
+  return (Math.floor(((17 * Math.PI) / 8 + angle) / (Math.PI / 4)) + idx) % 8;
```
Comment (targos) [inline src/utilities/corners.ts:17]: Do you have some notes about this formula? Can you put them in a comment?

**Comment by stropitek** [comment]
Thanks for the feedback!
> 1. I would not show the rotation handle while the ROI is being created

Fixed

> 2\. I'm able to make it go out of the boundaries:

Fixed

> 3\. The rotate handle can be brought outside of the boundaries, making it impossible to see/use. Maybe we can accept this limitation.

I'll create a new issue for this.


> 4\. Did you consider drawing a line between the edge of the rectangle and the circle?

I added the line. WDYT? I don't really have a preference.

File: src/utilities/corners.ts (hunk 1 above)
Comment (stropitek) [reply src/utilities/corners.ts:21]: Added an explanation. SGTY?

Comment (stropitek) [reply src/utilities/corners.ts:21]: https://github.com/zakodium-oss/react-roi/pull/122/commits/04c23bcfb03c5c61b0f66faab76a412ac8f98acd

Comment (targos) [reply src/utilities/corners.ts:21]: SGTM

**Comment by targos** [comment]
> I added the line. WDYT? I don't really have a preference.

I prefer with the line.

**Review by targos — APPROVED** [review]
I gave only a quick look to the code changes.

**Pull request merged**